src/connection.rs

```diff
@@ -622,3 +621,243 @@ pub fn transaction<K: ToRedisArgs,
+            ConnectionAddr::Tcp(ref host, ref port) => {
+                let host: &str = &*host;
+                // FIXME remove unwrap, make host IpAddr before this point?
+                Either::A(
+                    TcpStream::connect(&SocketAddr::new(host.parse().unwrap(), *port), handle)
+                      .from_err()
+                      .map(|con| ActualConnection::Tcp(BufReader::new(con)))
+                )
+            }
+            #[cfg(feature="with-unix-sockets")]
+            ConnectionAddr::Unix(ref path) => {
+                let result = UnixStream::connect(path, handle).map(|stream| {
+                    ActualConnection::Unix(BufReader::new(stream))
+                });
+                Either::B(future::result(result))
+            }
+            #[cfg(not(feature="with-unix-sockets"))]
+            ConnectionAddr::Unix(_) => {
+                Either::B(future::err(RedisError::from((ErrorKind::InvalidClientConfig,
+                       "Cannot connect to unix sockets \
+                       on this platform"))))
+            }
+        };
+
+        Box::new(connection.from_err().and_then(move |con| {
+            let rv = Connection {
+                con,
+                db: connection_info.db,
+            };
+
+            let login = match connection_info.passwd {
+                Some(ref passwd) => {
+                    Either::A(cmd("AUTH").arg(&**passwd).query_async::<_, Value>(rv)
+                        .then(|x| match x {
+                                Ok((rv, Value::Okay)) => Ok(rv),
+                                _ => {
+                                    fail!((
+                                        ErrorKind::AuthenticationFailed,
+                                        "Password authentication failed"
+                                    ));
+                                }
+                            })
+                        )
+                },
+                None => Either::B(future::ok(rv))
+            };
+
+            login.and_then(move |rv| {
+                if connection_info.db != 0 {
+                    Either::A(cmd("SELECT").arg(connection_info.db).query_async::<_, Value>(rv)
+                        .then(|result| match result {
+                                Ok((rv, Value::Okay)) => Ok(rv),
+                                _ => fail!((
+                                    ErrorKind::ResponseError,
+                                    "Redis server refused to switch database"
+                                )),
+                            })
+                        )
+                } else {
+                    Either::B(future::ok(rv))
+                }
+            })
+        }))
+    }
+
+    pub trait ConnectionLike: Sized {
+        /// Sends an already encoded (packed) command into the TCP socket and
+        /// reads the single response from it.
+        fn req_packed_command(self, cmd: Vec<u8>) -> RedisFuture<(Self, Value)>;
```

Comment:
I may have stuck a bit to close to the synchronous versions by using this pattern actually. I will be making a separate PR (building on this one) hopefully today where the `Connection` allows multiple commands to be in flight concurrently without needing an explicit pipeline. The trade off is that it won't be able to handle explicitly built pipelines (at least not initially) and there is some additional overhead due to needing `Arc`/`Rc`.